src/sessions.ts

```diff
@@ -181,24 +174,34 @@ export class ClientSession extends TypedEventEmitter<ClientSessionEvents> {
     this.clusterTime = options.initialClusterTime;
 
     this.operationTime = undefined;
-    this.explicit = !!options.explicit;
     this.owner = options.owner;
     this.defaultTransactionOptions = Object.assign({}, options.defaultTransactionOptions);
     this.transaction = new Transaction();
   }
 
   /** The server id associated with this session */
   get id(): ServerSessionId | undefined {
-    return this.serverSession?.id;
+    const serverSession = this[kServerSession];
+    if (serverSession == null) {
+      return undefined;
+    }
+    return serverSession.id;
   }
 
   get serverSession(): ServerSession {
-    if (this[kServerSession] == null) {
-      this[kServerSession] = this.sessionPool.acquire();
+    if (this.hasEnded) {
+      // @ts-expect-error: If the session has ended we do not want to run the acquire code below
+      // regardless of the value of kServerSession potentially being nullish. It *should* always be
+      // a ServerSession at this stage, but if it is not, risking a null access seems worth it
+      // as opposed to accidentally acquiring a new ServerSession for an ended ClientSession
+      return this[kServerSession];
```

Comment:
If we are confident that `this[kServerSession]` cannot be undefined here, then we can safely add a `if !this[kServerSession] { throw new MongoRuntimeError('...') }`. We shouldn't be using ts expect error in production code.

As for nullish values in general, you didn't really answer my question about why we can't initialize `this[kServerSession]` to null instead of undefined  and avoid any need for `ServerSession | undefined` typings altogether.

Reply:
Added throw